### PR TITLE
Support using comments to select parts to encrypt

### PR DESCRIPTION
**Pull request by felixfontein**
This PR adds support to annotate comments with a string (e.g., `sops:enc`) which can then be matched with a regex. If it matches, the corresponding value (the one which follows the comment) is encrypted while other values are not. (There is also the opposite regex available, to select those values which should not be encrypted.)

This enables the YAML file to have the same structure encrypted and decrypted, without having to add suffixes or manage complex regexes to match keys. See #543 for more discussion.

~~(This PR currently contains #1389 and will be rebased once that's merged. It is a lot more likely that #1389 will be merged before this PR or any of its variants, so I decided to use that one as a basis.)~~

This PR continues the work of #974 by rebasing #974 upon #1389 and adding some final touches (see https://github.com/getsops/sops/pull/974#issuecomment-1862766339).

Closes #974.

### CODE REVIEW

**Comment by felixfontein** [comment]
@mitar I think the two comments I added address the issues from my review in https://github.com/getsops/sops/pull/974#pullrequestreview-1642951900. Can you take a look at the commits?

File: sops.go
Comment (mitar) [inline sops.go:447]: Why not use here `MatchString`?

Didn't we [say we will also try to match against `EncryptedCommentRegex`](https://github.com/getsops/sops/pull/974#issuecomment-1793750862)? So both of them should not match the resulting encrypted string?

**Comment by mitar** [comment]
Thanks. I think this looks good as it is already. Thank you for all the work. I made a small comment, but I think it is not critical to do it.

File: sops.go
Comment (felixfontein) [reply sops.go:447]: I originally implemented this, and then when trying to test it I removed the code again. As opposed to `UnencryptedCommentRegex`, encrypted comments matching `EncryptedCommentRegex` aren't a problem since comments are only encrypted/decrypted if they are part of an encrypted subtree that's encrypted due to a comment higher up.

For example, you can encrypt
```yaml
---
#ENC[AES256_GCM,data:MI3A,iv:c9lHpZCbz3W9ooGuGbmQTcWcXAH7JEL/5JyJdC41dec=,tag:LrQLBEgQayguId5RrNiOCA==,type:comment]
foo:
  bar: barbar #ENC[AES256_GCM,data:MI3A,iv:c9lHpZCbz3W9ooGuGbmQTcWcXAH7JEL/5JyJdC41dec=,tag:LrQLBEgQayguId5RrNiOCA==,type:comment]
# something else
baz:
  bam: bambam #ENC[AES256_GCM,data:MI3A,iv:c9lHpZCbz3W9ooGuGbmQTcWcXAH7JEL/5JyJdC41dec=,tag:LrQLBEgQayguId5RrNiOCA==,type:comment]
```
with `--encrypted-comment-regex ENC` and then decrypt it again without any problems.

We could still implement the check to be extra cautious, but it isn't needed (or I'm missing something :) ).

Comment (mitar) [reply sops.go:447]: OK. I think it is then fine and we can revisit this if anyone finds a realistic case where it is breaking some user flow.

**Comment by felixfontein** [comment]
Now that #1389 has been merged, this is ready for review! :tada:

File: sops.go
Comment (felixfontein) [reply sops.go:266]: That can't happen since in the first line of the function, one element is appended to `commentsStack`, and during the execution of this function no value is removed from `commentsStack`.

**Comment by Gui13** [comment]
This looks good, any idea when this could be merged ?

**Comment by felixfontein** [comment]
It's currently waiting for further reviews / approval by maintainer(s).

**Comment by mitar** [comment]
Is there anything to do to push this further?

**Comment by felixfontein** [comment]
Like several other PRs, this is waiting for a review / approval from someone else in @getsops/maintainers.

**Review by devstein — COMMENTED** [review]
Looks good to me, but would be nice to get another review from @hiddeco @onedr0p if possible; otherwise, I can merge

**Review by sabre1041 — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@mitar thank you very much for your contribution! I'm glad it finally got merged...
Also thanks to everyone who helped reviewing and testing this!

**Comment by mitar** [comment]
Thank you everyone who was involved into getting this in, and especially you @felixfontein for all the help getting it in.

How could I bring #1404 further?

**Comment by c33s** [comment]
@felixfontein 

thank you very much for this awesome feature!

two questions: 
- can you please add a simple example to the readme on how to use this new feature?
- is it possible to invert the features behavior (encrypt by default and use a comment to keep the value unencrypted)?

**Comment by mitar** [comment]
README does have some text about this [here](https://github.com/getsops/sops/pull/1392/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168f).

> is it possible to invert the features behavior (encrypt by default and use a comment to keep the value unencrypted)?

Yes, you can use `unencrypted-regex` for that.

**Comment by c33s** [comment]
i meant some examples for the config file like:
.sops.yaml
```
- encrypted_regex: '.'
  unencrypted_comment_regex: 'keep_decrypted'
  path_regex: '*\.yaml$'
```

file.yaml
```
foo:
  # keep_decrypted
  bar: 1234
  password: will be enrcypted
```

oh in missed the `--unencrypted-comment-regex`. really awesome it can work both ways

**Pull request merged**
